### ISSUE TITLE
Indexing performance improvements

### ISSUE THREAD

**Issue by Jannis** [graphprotocol/graph-node#1347]
Let's gather information about indexing performance and create a list of the 4-5 highest impact changes we think we can make to boost indexing speed.

**Comment by leoyvens** [comment]
The following data is from a local run of ERC20, for 100 minutes (so we can read the minutes as percentages 😃), which got it to block ~4729000:
![Captura de Tela 2019-11-19 às 10 35 22](https://user-images.githubusercontent.com/9885558/69150884-5ab08700-0ab8-11ea-86cb-bfb951f1e816.png)

Caveats on this data: It's from my local machine, no other subgraphs deployed, DB with blocks cached, single run, and it's not even close to a full sync. This needs to be confirmed with data from a longer run on a production server. Still, I feel like some trends should be representative and we can start thinking about solutions.

1. `store.get` and `store.set` combined accounted for ~35% of time spent. Both require a trip to the DB to find the entity with the specified id. I measured that 90% of these calls are for non-existing entities, on erc20 that's probably because it's common practice to create new accounts for each token transfer. Keeping an in-memory bloom filter would allow us to skip the DB on these cases, and if this proportion holds up through the rest of the blocks, it would be a 30% improvement to indexing time!

2. Inserts and updates in `apply_entity_modifications` turn out to be the most expensive part of `transact_block`, which is not surprising, and take ~30% of indexing time. I tried disabling attribute indexes, and did not measure any improvement, maybe as more entities as inserted that becomes noticeable. So it could be that the time spent is due to the overhead of doing a DB call for each operation. If we speculate that the overhead of each DB roundtrip is in the order of 1ms, doing one per operation is significant. This could be improved by using `COPY` instead of `INSERT`, and any other tricks we can think of to reduce the number of calls.

3. `scan_blocks` took ~14% of the time. We'd need to break that down further to find bottlenecks, but the actual `eth_getLogs` calls are likely it. #1350 might change the picture here, not sure if for better or worse.

4. `run_handler` took ~8% of the time. This is time spent actually executing WASM. Parallelizing handler execution is one idea. Another is switching to compiled WASM execution.

Pipelining is also an idea, for example scanning blocks in the background and queuing the triggers to be executed, and/or queuing the entity operations to be transacted into the DB. This is architecturally tricky to do, but if we end up with two sections run sequentially and take roughly the same time to execute, pipelining can be a 2x gain.
